src/net.cpp

```diff
@@ -921,11 +921,13 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
         {{false, NET_I2P, 0}, {/* localhost */ true, NET_MAX, 0}, {false, NET_ONION, 0}}};
 
     // Count and store the number of eviction candidates per network.
-    for (Net& n : networks) {
-        n.count = std::count_if(eviction_candidates.cbegin(), eviction_candidates.cend(),
-                                [&n](const NodeEvictionCandidate& c) {
-                                    return n.is_local ? c.m_is_local : c.m_network == n.id;
-                                });
+    for (auto& c : eviction_candidates) {
+        for (int i{networks.size() - 1}; i >= 0; --i) {
+            if (networks[i].is_local ? c.m_is_local : c.m_network == networks[i].id) {
+                ++networks[i].count;
+                break;
+            }
+        }
     }
```

Comment:
99a25fb393 `p2p: iterate over eviction candidates once instead of thrice`

I did not observe any difference in performance due to this commit.

It changes the complexity from `O(num networks * num candidates)` to `O(num candidates * num networks)`. The early `break` optimization maybe makes it `O(num candidates * num networks * 0.5)` on average, but that is equal to `O(num candidates * num networks)`.

Just to experiment I made this `O(num candidates)` which is as fast as it could be (unfortunately a bit hackish due to the hardcoded indexes):

```diff
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -918,18 +918,26 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     // array members have first opportunity to recover unused slots from the previous iteration.
     struct Net { bool is_local; Network id; size_t count; };
     std::array<Net, 3> networks{
         {{false, NET_I2P, 0}, {/* localhost */ true, NET_MAX, 0}, {false, NET_ONION, 0}}};
 
     // Count and store the number of eviction candidates per network.
-    for (auto& c : eviction_candidates) {
-        for (int i{networks.size() - 1}; i >= 0; --i) {
-            if (networks[i].is_local ? c.m_is_local : c.m_network == networks[i].id) {
-                ++networks[i].count;
-                break;
-            }
+    for (const auto& c : eviction_candidates) {
+        const Network net = c.m_is_local ? NET_MAX : c.m_network;
+        switch (net) {
+        case NET_I2P:
+            ++networks[0].count;
+            break;
+        case NET_MAX:
+            ++networks[1].count;
+            break;
+        case NET_ONION:
+            ++networks[2].count;
+            break;
+        default:
+            break;
         }
     }
     // Sort `networks` by ascending candidate count, to give networks having fewer candidates
     // the first opportunity to recover unused protected slots from the previous iteration.
```

No difference in performance either. I guess this is operating on too small amount of data to make any performance impact and so this commit can be dropped if its purpose is to improve performance and we can't observe any changes.

Reply:
> I did not observe any difference in performance due to this commit.

Here are the results I get:

272f327fbf `bench: add peer eviction protection benchmarks`
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      706,673,971.00 |                1.42 |    1.7% |      7.79 | `EvictionProtection1Network250Candidates`
|    1,299,063,388.00 |                0.77 |    0.7% |     14.30 | `EvictionProtection3Networks050Candidates`
|    1,155,171,747.00 |                0.87 |    0.5% |     12.73 | `EvictionProtection3Networks100Candidates`
|    2,900,919,398.00 |                0.34 |    1.3% |     31.93 | `EvictionProtection3Networks250Candidates`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      714,684,192.00 |                1.40 |    1.0% |      7.91 | `EvictionProtection1Network250Candidates`
|    1,296,579,234.00 |                0.77 |    0.5% |     14.22 | `EvictionProtection3Networks050Candidates`
|    1,159,393,440.00 |                0.86 |    0.7% |     12.71 | `EvictionProtection3Networks100Candidates`
|    2,854,527,930.00 |                0.35 |    0.7% |     31.44 | `EvictionProtection3Networks250Candidates`
```

99a25fb393 `p2p: iterate over eviction candidates once instead of thrice`
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      709,622,212.00 |                1.41 |    1.0% |      7.79 | `EvictionProtection1Network250Candidates`
|    1,310,376,054.00 |                0.76 |    1.2% |     14.39 | `EvictionProtection3Networks050Candidates`
|    1,160,800,660.00 |                0.86 |    0.4% |     12.72 | `EvictionProtection3Networks100Candidates`
|    2,983,258,845.00 |                0.34 |    0.5% |     32.71 | `EvictionProtection3Networks250Candidates`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      711,091,374.00 |                1.41 |    0.5% |      7.82 | `EvictionProtection1Network250Candidates`
|    1,318,781,548.00 |                0.76 |    0.3% |     14.38 | `EvictionProtection3Networks050Candidates`
|    1,142,837,118.00 |                0.88 |    0.2% |     12.40 | `EvictionProtection3Networks100Candidates`
|    2,943,154,182.00 |                0.34 |    0.4% |     32.22 | `EvictionProtection3Networks250Candidates`
```

Reply:
99a25fb393c791bb6b6edeab3677d39bf6b1d97d Yes, ISTM only the first test would utilize the early break and (potentially) show any difference. I agree that it's probably not worth it.

Reply:
99a25fb393c791bb6b6edeab3677d39bf6b1d97d I agree that it's probably not worth it.

Reply:
Dropped.